pkg/relayer/session/proof.go

```diff
@@ -0,0 +1,111 @@
+		ctx, claimedSessions,
+		rs.mapWaitForEarliestSubmitProofHeight,
+	)
+
+	failedSubmitProofSessions, failedSubmitProveSessionsPublishCh :=
+		channel.NewObservable[relayer.SessionTree]()
+
+	// Map sessionsWithOpenProofWindow to a new observable of an either type,
+	// populated with the session or an error, which is notified after the session
+	// proof has been submitted or an error has been encountered, respectively.
+	eitherProvenSessions := channel.Map(
+		ctx, sessionsWithOpenProofWindow,
+		rs.newMapProveSessionFn(failedSubmitProveSessionsPublishCh),
+	)
+
+	// TODO_TECHDEBT: pass failed submit proof sessions to some retry mechanism.
+	_ = failedSubmitProofSessions
+	logging.LogErrors(ctx, filter.EitherError(ctx, eitherProvenSessions))
+}
+
+// mapWaitForEarliestSubmitProofHeight is intended to be used as a MapFn. It
+// calculates and waits for the earliest block height, allowed by the protocol,
+// at which a proof can be submitted for the given session, then emits the session
+// **at that moment**.
+func (rs *relayerSessionsManager) mapWaitForEarliestSubmitProofHeight(
+	ctx context.Context,
+	session relayer.SessionTree,
+) (_ relayer.SessionTree, skip bool) {
+	rs.waitForEarliestSubmitProofHeight(
+		ctx, session.GetSessionHeader().GetSessionEndBlockHeight(),
+	)
+	return session, false
+}
+
+// waitForEarliestSubmitProofHeight calculates and waits for (blocking until) the
+// earliest block height, allowed by the protocol, at which a proof can be submitted
+// for a session which was claimed at createClaimHeight. It is calculated relative
+// to createClaimHeight using on-chain governance parameters and randomized input.
+func (rs *relayerSessionsManager) waitForEarliestSubmitProofHeight(
+	ctx context.Context,
+	createClaimHeight int64,
+) {
+	submitProofWindowStartHeight := createClaimHeight
+	// TODO_TECHDEBT: query the on-chain governance parameter once available.
+	// + claimproofparams.GovSubmitProofWindowStartHeightOffset
+
+	// we wait for submitProofWindowStartHeight to be received before proceeding since we need its hash
+	log.Printf("waiting for global earliest proof submission submitProofWindowStartBlock height: %d", submitProofWindowStartHeight)
+	submitProofWindowStartBlock := rs.waitForBlock(ctx, submitProofWindowStartHeight)
+
+	earliestSubmitProofHeight := protocol.GetEarliestSubmitProofHeight(submitProofWindowStartBlock)
+	_ = rs.waitForBlock(ctx, earliestSubmitProofHeight)
+}
+
+// newMapProveSessionFn returns a new MapFn that submits a proof for the given
+// session. Any session which encouters errors while submitting a proof is sent
+// on the failedSubmitProofSessions channel.
+func (rs *relayerSessionsManager) newMapProveSessionFn(
+	failedSubmitProofSessions chan<- relayer.SessionTree,
+) channel.MapFn[relayer.SessionTree, either.SessionTree] {
+	return func(
+		ctx context.Context,
+		session relayer.SessionTree,
+	) (_ either.SessionTree, skip bool) {
+		latestBlock := rs.blockClient.LatestBlock(ctx)
```

Comment:
```suggestion
		// TODO_BLOCKER: The block that'll be used as a source of entropy for which 
		// branch(es) to prove should be deterministic and use on-chain governance params
		// rather than latest.
		latestBlock := rs.blockClient.LatestBlock(ctx)
```